docs: Enhance GitHub Sign-In setup, customizations, and troubleshooting (#521)
serverpod/serverpod_docs: docs/06-concepts/11-authentication/04-providers/07-github/01-setup.md

# Setup

Sign in with GitHub uses OAuth2 credentials from a **GitHub App** registered on GitHub.

## Prerequisites

Before following this guide, make sure you have:

- A GitHub account with access to [Developer Settings](https://github.com/settings/apps).
- A running Serverpod project (server, client, and Flutter app packages from `serverpod create`).
- The Serverpod auth module installed and configured per the [authentication setup](../../setup). If your project was generated with an older Serverpod version, follow that guide first to add `serverpod_auth_idp_server` and `serverpod_auth_idp_flutter` and to configure `pod.initializeAuthServices()` before continuing.

:::note

> If you specifically need an **OAuth App** instead of a GitHub App, the setup flow below is the same except OAuth Apps allow only a single Callback URL (GitHub Apps allow up to 10). See GitHub's [Differences between GitHub Apps and OAuth Apps](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/differences-between-github-apps-and-oauth-apps).

Reviewer: This note will just confuse most users.

- It requires too much pre-existing knowledge about GitHub´s ecosystem. 
- It mentions technical limitations, but not why the user should care. 
- It does not tell why to choose one over another, or what options is better for standard Serverpod setup.

Just remove the note.

Author: Removed

:::

## Get your GitHub credentials

### Register a new GitHub App

1. Go to [Register new GitHub App](https://github.com/settings/apps/new).

2. Fill in the basics:

>    - **GitHub App name** (required, up to 34 characters, unique across GitHub).

Reviewer: Leave form validation for GitHub to explain.

```suggestion
   - **GitHub App name**.
```

Author: Done

>    - **Homepage URL** (required).

Reviewer: ```suggestion
   - **Homepage URL**
```

>    - **Description** (optional, shown to users during install).

Reviewer: ```suggestion
   - **Description** (optional).
```

Author: Done

![GitHub App basics](/img/authentication/providers/github/1-app-basics.png)

### Configure the callback URL

> The callback URL is where GitHub redirects the user after they authorize your app. For Serverpod sign-in, this must match the redirect URI you pass to `initializeGitHubSignIn` in your Flutter app, and the registered URL in the web callback page setup.

Reviewer: These terms are not familiar to the user yet, it comes later. It should instead be mentioned when setting up the URI/URL for the Flutter app and callback page, that it must reflect the callback URL they entered in the GitHub portal.

```suggestion
The callback URL is where GitHub redirects the user after they authorize your app.
```

Author: Done

1. In the **Callback URL** field, enter the redirect URI for your app. GitHub Apps accept up to 10 entries, one per line. Add every platform you target:

>    | Platform | Example value |

Reviewer: The table is quite verbose. Can we make it into a bullet list to enhance readability?

Author: Done, converted to bullets

   | --- | --- |
   | iOS and Android | `com.example.yourapp://auth` (custom scheme registered in `AndroidManifest.xml` and `Info.plist`) |
   | Web (Serverpod-hosted Flutter) | `https://my-awesome-project.serverpod.space/auth/callback` |
   | Web (Flutter dev server) | `http://localhost:49660/auth.html` |

>    The web entries depend on how your Flutter web app is served. See [Web](#web) below for the two flows (Serverpod-hosted using a built-in callback route, or separately-hosted using a static `auth.html` file).

Reviewer: This sentence forces the user to leave the flow of the page and start researching below.

To solve this we need to stick to one default path throughout the guide. The default path is to set it up locally, and have the web app served by Serverpod in production. 

Hosting it seperately is an advanced use case.

What complicates this is when running it locally it is essentially hosted separately. However, we need to speak with the use cases in mind. IE, the user want to test it locally, not host it seperately, so that's the language we should use.

Then have a section under customizations about hosting the app separately. (While also avoiding duplicate information).

Author: restructured to match the google setup. default path is now serverpod-hosted for local + prod, moved the separately-hosted flow to customizations

   :::tip

>    For the mobile scheme, use a value unique to your app (reverse-DNS of your bundle ID is a good convention, for example `com.example.yourapp://auth`). Generic schemes like `myapp://auth` work but can collide with other apps installed on the device.

Reviewer: This tip is not related to setting up GitHub authentication. It's just noise. Remove it.

Author: Removed

   :::

   ![Callback URL field](/img/authentication/providers/github/2-callback-url.png)

> 2. Leave **Expire user authorization tokens** enabled (GitHub's default). Token expiration is recommended for sign-in flows so leaked tokens have a short useful lifetime. Serverpod handles the refresh flow automatically; you do not need to write any refresh logic.

Reviewer: ```suggestion
2. Leave **Expire user authorization tokens** enabled (GitHub's default). Token expiration is recommended for sign-in flows so leaked tokens have a short useful lifetime. Serverpod handles refreshing the token; you do not need to write any refresh logic.
```

Author: Done

> 3. Leave **Request user authorization (OAuth) during installation** unchecked unless you need the installation to immediately trigger an OAuth sign-in.

Reviewer: ```suggestion
3. Leave **Request user authorization (OAuth) during installation** unchecked unless you need the installation of your Flutter app to immediately trigger an OAuth sign-in.
```

Author: Done

### Disable webhooks

Webhooks let your GitHub App receive events like pushes, pull requests, and issue activity. They are unrelated to sign-in and add complexity if you do not need them, so turn the section off for now. You can re-enable webhooks later without affecting the auth flow.

Under **Webhook**, uncheck **Active** to disable the whole section.

![Disable webhook](/img/authentication/providers/github/3-webhook-disable.png)

### Set permissions

> GitHub Apps use fine-grained **permissions** instead of OAuth **scopes**. For sign-in you only need access to the user's profile.

Reviewer: Remove the technical noise. Make it simple.

```suggestion
For sign-in you only need access to the user's profile.
```

Author: Done

### Web

On web, GitHub completes sign-in by redirecting the browser to a callback URL you control. This flow requires Serverpod to serve your Flutter web app on the **same origin** as the callback route. To test locally, build your Flutter web app into Serverpod's `web/app/` directory:

```bash
flutter build web --output ../my_project_server/web/app  # from your Flutter project

> dart run bin/main.dart                                   # from your server project

Reviewer: Is this the recommended workflow for this even though we have the new serverpod start workflow?

Author: Switched to `serverpod start --no-flutter`. Plain `serverpod start` also launches `flutter run -d chrome` which won't share Serverpod's origin, so the flag keeps it serving the prebuilt app at /app. Tested locally on beta.8.

1. In your server project's `server.dart` file, before `pod.start()`, register the callback route:

   ```dart
   import 'package:serverpod_auth_idp_server/core.dart';

   // ...

   pod.webServer.addRoute(
     FlutterWebAuth2CallbackRoute(),
     '/auth/callback',
   );
   ```

   The path `/auth/callback` can be anything, as long as it matches the URL you register with GitHub in step 2 and pass to `initializeGitHubSignIn` later.

   In production, pass `host:` to restrict the route to your production domain:

   ```dart
   pod.webServer.addRoute(
     FlutterWebAuth2CallbackRoute(host: 'my-awesome-project.serverpod.space'),
     '/auth/callback',
   );
   ```

   :::tip
   The same route works for every OAuth2 PKCE provider. Register it once and reuse the callback URL for GitHub, Google, and any future provider.
   :::

2. Add the full callback URL to your GitHub App's **Callback URL** field (e.g., `http://localhost:8082/auth/callback` locally, `https://my-awesome-project.serverpod.space/auth/callback` in production). GitHub Apps accept up to 10 entries, so dev and prod entries can coexist.

> 3. Pass the same URL to `initializeGitHubSignIn` via the `redirectUri` argument when you initialize the client (covered in [Initialize the GitHub sign-in service](#initialize-the-github-sign-in-service) below).

Reviewer: Having a step here mentioning something below gets confusing. Instead, refer back to the callback URL in step 2, from the section below, and remove this step 3. It adds noise and confusion.

Author: Removed step 3. The init section below already points back to the callback URL.

   :::tip
   You can also pass the redirect URI via `--dart-define`. See [Configuring client IDs on the app](./customizations#configuring-client-ids-on-the-app) for the pattern.
   :::

:::note

> `FlutterWebAuth2CallbackRoute` requires `serverpod_auth_idp_server` 3.5.0-beta.8 or later.

Reviewer: Requirements shall be put at the top before starting, just as in the other guides.

Author: Moved to the prerequisites at the top.
